### PR TITLE
core: fix cookie parsing (part 2)

### PR DESCRIPTION
**Pull request by ngosang**
After fixing cookie storage in #8133 I noticed that I still have a lot of '.json.bak' files in the Jackett configuration folder. After deleting them they were created again in each request. The cause was we were parsing bad the cookies with '=' character in the value. Most Cloudflare cookies include it so we were sending bad cookies and solving the challenge in each request.

This PR should increase performance in several ways: we are not solving the challenge again (it takes time), we are not making extra requests and we are not updating the Jackett configuration in each request (both files '.json' and '.json.bak').

Tested with the client HttpWebClient2NetCore only. Please do some tests with the site 1337x in other clients.

### CODE REVIEW

File: src/Jackett.Common/Utils/CookieUtil.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            return cookieDictionary;
+        }
+
+        public static string CookieDictionaryToHeader(Dictionary<string, string> cookieDictionary)
+        {
+            if (cookieDictionary == null)
+                return "";
+            return string.Join(
+                "; ", cookieDictionary
+                      .Where(kv => kv.Key.IndexOfAny(InvalidKeyChars) < 0 && kv.Value.IndexOfAny(InvalidValueChars) < 0 )
```
Comment (cadatoiva) [inline src/Jackett.Common/Utils/CookieUtil.cs:35]: Based on your given test cases below, maybe completely skipping malformed keys/values might not be the best route? I feel like maybe web encoding or truncating would potentially be better. Just looking through your test cases, I would have thought `{"__cf_bm", "34234 234"}` would become either `"__cf_bm=34234"` or `"__cf_bm=34234+234"` since these dictionaries would be built either by your conversion code, or by devs themselves, it feels like `Dictionary->string` would be more on purpose with choice of key/value.

File: src/Jackett.Common/Utils/CookieUtil.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+using System.Text.RegularExpressions;
+
+namespace Jackett.Common.Utils
+{
+    public static class CookieUtil
+    {
+        // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie
+        private static readonly Regex _CookieRegex = new Regex(@"([^\(\)<>@,;:\\""/\[\]\?=\{\}\s]+)=([^,;\\""\s]+)");
+        private static readonly char[] InvalidKeyChars = {'(', ')', '<', '>', '@', ',', ';', ':', '\\', '"', '/', '[', ']', '?', '=', '{', '}', ' ', '\t', '\n'};
+        private static readonly char[] InvalidValueChars = {'"', ',', ';', '\\', ' ', '\t', '\n'};
```
[the remark this answers is not in the document]
Comment (ngosang) [reply src/Jackett.Common/Utils/CookieUtil.cs:12]: The cookie format was defined long time ago and I don't think it will change in the future. I prefer to keep the code as is.

File: src/Jackett.Common/Utils/CookieUtil.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+        public static Dictionary<string, string> CookieHeaderToDictionary(string cookieHeader)
```
[the remark this answers is not in the document]
Comment (ngosang) [reply src/Jackett.Common/Utils/CookieUtil.cs:14]: I don't think extension methods are for this use case. The 99.9% of the strings we are using don't contains cookies and calling this method will generate an empty dictionary. Why would I want an useless methods in all strings? I think extension methods are for more generic things like "special join" or something that could be apply to all strings.

File: src/Jackett.Common/Utils/CookieUtil.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            while (matches.Success)
+            {
+                if (matches.Groups.Count > 2)
+                    cookieDictionary[matches.Groups[1].Value] = matches.Groups[2].Value;
+                matches = matches.NextMatch();
+            }
+            return cookieDictionary;
+        }
+
+        public static string CookieDictionaryToHeader(Dictionary<string, string> cookieDictionary)
```
[the remark this answers is not in the document]
Comment (ngosang) [reply src/Jackett.Common/Utils/CookieUtil.cs:29]: no, sorry

File: src/Jackett.Common/Utils/CookieUtil.cs (hunk 1 above)
Comment (ngosang) [reply src/Jackett.Common/Utils/CookieUtil.cs:35]: Well, I choose a silent approach because the previous code never fails. Changing the behavior can cause problems if some indexer is wrong.
I'm going to change it to raise an exception instead, the user will be responsible for encoding the value properly. Most of the cookies we use come from webservers and they are right.

Comment (ngosang) [reply src/Jackett.Common/Utils/CookieUtil.cs]: Changed, I tested all public trackers and is fine throwing the exception.

File: src/Jackett.Common/Utils/CookieUtil.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (cadatoiva) [reply src/Jackett.Common/Utils/CookieUtil.cs:13]: The link you shared seems to be updated pretty regularly, with 4 updates in 2020 according to their diffs. The current cookie-standard is 9 years old this month, with 3 major updates between 1997 and 2011, each of which changed what counts as a valid character according to https://stackoverflow.com/a/1969339.

It's also come to my attention that the whole Regex is incomplete, seeing as non-ASCII characters are also invalid in full. Seems like a whitelist of allowed chars might be better?

There's also the whole mess that many sites may pass any of the RFC specs or even the original Netscape standard for the cookies (according to that stack overflow link), meaning that the websites could be passing and expecting back cookies this code removes as invalid.

File: src/Jackett.Common/Utils/CookieUtil.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (cadatoiva) [reply src/Jackett.Common/Utils/CookieUtil.cs:15]: I agree.

File: src/Jackett.Common/Utils/CookieUtil.cs (hunk 1 above)
Comment (cadatoiva) [reply src/Jackett.Common/Utils/CookieUtil.cs]: That works

File: src/Jackett.Test/Utils/CookieUtilTests.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            var cookieDictionary = new Dictionary<string, string> {{"__cf_=bm", "34234234"}};
+            try
+            {
+                CookieUtil.CookieDictionaryToHeader(cookieDictionary);
+                Assert.Fail("Malformed cookie must fail");
+            }
+            catch (FormatException e)
+            {
+                Assert.AreEqual("The cookie '__cf_=bm=34234234' is malformed.", e.Message);
+            }
```
Comment (cadatoiva) [inline src/Jackett.Test/Utils/CookieUtilTests.cs:85]:
```suggestion
            Assert.Throws<FormatException>(
                () => CookieUtil.CookieDictionaryToHeader(cookieDictionary), "The cookie '__cf_=bm=34234234' is malformed.");
```

File: src/Jackett.Test/Utils/CookieUtilTests.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+            };
+            try
+            {
+                CookieUtil.CookieDictionaryToHeader(cookieDictionary);
+                Assert.Fail("Malformed cookie must fail");
+            }
+            catch (FormatException e)
+            {
+                Assert.AreEqual("The cookie '__cf_bm=34234 234' is malformed.", e.Message);
+            }
```
Comment (cadatoiva) [inline src/Jackett.Test/Utils/CookieUtilTests.cs:104]:
```suggestion
            Assert.Throws<FormatException>(
                () => CookieUtil.CookieDictionaryToHeader(cookieDictionary), "The cookie '__cf_bm=34234 234' is malformed.");
```

File: src/Jackett.Common/Utils/CookieUtil.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ngosang) [reply src/Jackett.Common/Utils/CookieUtil.cs:13]: I agree we can add no-ascii checks but I'm worried that that regex doesn't include all valid characters and causes problems.

I'm happy with the current implementation, it fix the bug, has unit tests, has been tested by me with almost 300 public sites. This code is much better than the previous one so I don't see any reason to not merge.

[the remark this answers is not in the document]
Comment (cadatoiva) [reply src/Jackett.Common/Utils/CookieUtil.cs:13]: Just want to make sure we get the parsing right, because we don't need weird bugs coming from this code. I think my suggestion is easier to follow and verify for anyone looking over this code in the future. And after looking into it further, it feels like this PR in general might introduce subtle bugs into some trackers.

Honestly, though, you've got more experience in web technologies, so I'll defer to your judgement. Just wanted to share my concerns in case they are valid.

[the remark this answers is not in the document]
Comment (ngosang) [reply src/Jackett.Common/Utils/CookieUtil.cs:14]: I appreciate your feedback, but sometimes I try to fix one small problem and it becomes solving many other problems at once with a bunch of messages in the middle.
I have added a comment for the future. I think it's fine the way it is.

File: src/Jackett.Test/Utils/CookieUtilTests.cs (hunk 5 above)
Comment (ngosang) [reply src/Jackett.Test/Utils/CookieUtilTests.cs]: You code is not working, it's not cheeking the exception message. The message in your code is the error shown if the assert fails. I did other thing.

File: src/Jackett.Common/Utils/CookieUtil.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (cadatoiva) [reply src/Jackett.Common/Utils/CookieUtil.cs:14]: Sometimes I feel the same way. Like adding >100 lines of Unit Tests, plus discussion on null handling, error handling, default values, and best practices spanning more than 50 comments back and forth over what was a 4 line change to a utility function in the GetQueryString PR. Frustrates me in the moment too, but I can see the importance of the discussion, the extra considerations, and planning the path forward in case we aren't completing the feature in a single PR.

File: src/Jackett.Test/Utils/CookieUtilTests.cs (hunk 5 above)
Comment (cadatoiva) [reply src/Jackett.Test/Utils/CookieUtilTests.cs]: My mistake, I just saw the intellisense called the parameter `message` and assumed it meant checking the message was equal.

**Comment by garfield69** [comment]
so you want me to test using some of the other clients. sure.
what is my criteria for success? what am I supposed to be looking for?

File: src/Jackett.Common/Utils/CookieUtil.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ngosang) [reply src/Jackett.Common/Utils/CookieUtil.cs:14]: I apologize for it. I just wanted you to understand the importance of testing, but now I realize that blocking PR was not the right way.

**Comment by ngosang** [comment]
@garfield69 I did all the test already. Just mention me if someone is having problems with cookies or cloudflare in the following days.

File: src/Jackett.Common/Utils/CookieUtil.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (cadatoiva) [reply src/Jackett.Common/Utils/CookieUtil.cs:14]: It's all good, I know it's not a malicious thing. And I want to clarify I wasn't delaying this PR out of spite or anything, it just came to mind when you vented your frustration. I do prefer the feedback and conversation we have on the PRs.

I want to point out that GitHub defaults to letting Jackett admins checkout and commit to open PR branches. So we can also implement the requested changes directly ourselves if need be, so that can be an option we take advantage of in future PRs if need be.
